File: tokyox/tools/terminal_tools.py
from __future__ import annotations
import asyncio
import os
import shlex
import sys
from typing import Any


def make_terminal_tool(cwd: str, defaults: dict[str, Any] | None = None):
    defaults = defaults or {}

    async def terminal_exec(args: dict[str, Any], ctx: dict[str, str]) -> dict[str, Any]:
        command = str(args.get("command", ""))
        timeout_ms = min(int(args.get("timeoutMs", defaults.get("timeoutMs", 15_000))), 60_000)
        shell = "cmd" if sys.platform == "win32" else "/bin/zsh"
        shell_args = ["/c", command] if sys.platform == "win32" else ["-lc", command]
        env = {**os.environ, "PATH": os.environ.get("PATH", ""), "HOME": os.environ.get("HOME", ""), "TOKYOX_ACTOR": ctx.get("actor", "")}

        proc = await asyncio.create_subprocess_exec(
            shell,
            *shell_args,
            cwd=cwd,
            env=env,
            stdout=asyncio.subprocess.PIPE,
            stderr=asyncio.subprocess.PIPE,
        )
        try:
            stdout, stderr = await asyncio.wait_for(proc.communicate(), timeout=timeout_ms / 1000)
        except asyncio.TimeoutError:
            proc.kill()
            await proc.wait()
            return {"code": -1, "signal": "SIGTERM", "stdout": "", "stderr": "", "timedOut": True, "truncated": True}
        cap = defaults.get("maxOutput", 100_000)
        out = stdout.decode(errors="replace")[:cap]
        err = stderr.decode(errors="replace")[:cap]
        return {
            "code": proc.returncode,
            "signal": None,
            "stdout": out,
            "stderr": err,
            "timedOut": False,
            "truncated": len(stdout) > cap or len(stderr) > cap,
        }

    return {"terminal.exec": terminal_exec}


TOOL_DEFINITIONS_TERMINAL: list = [
    # defined in tools/__init__.py
]

File: tokyo-x/src/tools/terminal-tools.ts
import { spawn } from "node:child_process";

export function createTerminalTool(cwd: string, defaults: { timeoutMs?: number; maxOutput?: number } = {}) {
  return {
    "terminal.exec": async (args: Record<string, unknown>, ctx: { actor: string }) => {
      const command = String(args.command ?? "");
      const timeoutMs = Math.min(Number(args.timeoutMs ?? defaults.timeoutMs ?? 15_000), 60_000);
      const shell = process.platform === "win32" ? "cmd" : "/bin/zsh";
      const shellArgs = process.platform === "win32" ? ["/c", command] : ["-lc", command];
      return await new Promise((resolve, reject) => {
        const child = spawn(shell, shellArgs, {
          cwd,
          env: { PATH: process.env.PATH, HOME: process.env.HOME, TOKYOX_ACTOR: ctx.actor },
          signal: AbortSignal.timeout(timeoutMs),
        });
        let out = "";
        let err = "";
        const cap = defaults.maxOutput ?? 100_000;
        child.stdout.on("data", (d) => {
          if (out.length < cap) out += d;
        });
        child.stderr.on("data", (d) => {
          if (err.length < cap) err += d;
        });
        child.on("error", reject);
        child.on("close", (code, sig) => {
          resolve({
            code,
            signal: sig,
            stdout: out.slice(0, cap),
            stderr: err.slice(0, cap),
            timedOut: sig === "SIGTERM",
            truncated: out.length >= cap || err.length >= cap,
          });
        });
      });
    },
  };
}